subsys/bluetooth/host/hci_core.c

```diff
@@ -1949,13 +1950,21 @@ int bt_unpair(uint8_t id, const bt_addr_le_t *addr)
 		return -EINVAL;
 	}
 
-	if (IS_ENABLED(CONFIG_BT_SMP) &&
-	    (!addr || bt_addr_le_eq(addr, BT_ADDR_LE_ANY))) {
-		bt_foreach_bond(id, unpair_remote, &id);
-		return 0;
+	if (IS_ENABLED(CONFIG_BT_SMP)) {
+		if (!addr || bt_addr_le_eq(addr, BT_ADDR_LE_ANY)) {
+			bt_foreach_bond(id, unpair_remote, &id);
+		} else {
+			unpair(id, addr);
+		}
+	} else {
+		CHECKIF(addr == NULL) {
+			LOG_DBG("addr is NULL");
+			return -EINVAL;
+		}
+
+		unpair(id, addr);
 	}
 
-	unpair(id, addr);
 	return 0;
```

Comment:
Looks complicated, isn't it that `CHECKIF` is missing only?

```
	if (IS_ENABLED(CONFIG_BT_SMP) &&
	    (!addr || bt_addr_le_eq(addr, BT_ADDR_LE_ANY))) {
		bt_foreach_bond(id, unpair_remote, &id);
		return 0;
	}

        CHECKIF(addr == NULL) {
	        LOG_DBG("addr is NULL");
	        return -EINVAL;
        }

	unpair(id, addr);
	return 0;
```

Reply:
It's a nitpick :)
The linker is smart enough to remove unused code, please check how it looks for you, but for me the output assembly is exactly the same (`CONFIG_BT_SMP=y`, `CONFIG_ASSERT=y`).